include/zim/archive.h

```diff
@@ -0,0 +1,424 @@
+       */
+      std::string getMetadata(const std::string& name) const;
+
+      /** Get the list of metadata stored in the archive.
+       *
+       *  @return The list of metadata in the archive.
+       */
+      std::vector<std::string> getMetadataKeys() const;
+
+      /** Get an entry using its "path" index.
+       *
+       *  Use the index of the entry to get the idx'th entry
+       *  (entry being sorted by path).
+       *
+       *  @param idx The index of the entry.
+       *  @return The Entry.
+       *  @exception std::out_of_range If idx is greater than the number of entry.
+       */
+      Entry getEntryByPath(entry_index_type idx) const;
+
+      /** Get an entry using a path.
+       *
+       *  Get an entry using its path.
+       *  The path must contains the namespace.
+       *
+       *  @param path The entry's path.
+       *  @return The Entry.
+       *  @exception EntryNotFound If no entry has the asked path.
+       */
+      Entry getEntryByPath(const std::string& path) const;
+
+      /** Get an entry using its "title" index.
+       *
+       *  Use the index of the entry to get the idx'th entry
+       *  (entry being sorted by title).
+       *
+       *  @param idx The index of the entry.
+       *  @return The Entry.
+       *  @exception std::out_of_range If idx is greater than the number of entry.
+       */
+      Entry getEntryByTitle(entry_index_type idx) const;
+
+      /** Get an entry using a title.
+       *
+       *  Get an entry using its path.
+       *
+       *  @param title The entry's title.
+       *  @return The Entry.
+       *  @exception EntryNotFound If no entry has the asked title.
+       */
+      Entry getEntryByTitle(const std::string& title) const;
+
+      /** Get an entry using its "cluster" index.
+       *
+       *  Use the index of the entry to get the idx'th entry
+       *  The actual order of the entry is not really specified.
+       *  It is infered from the internal way the entry are stored.
+       *
+       *  This method is probably not relevent and is provided for completeness.
+       *  You should probably use a iterator using the `efficientOrder`.
+       *
+       *  @param idx The index of the entry.
+       *  @return The Entry.
+       *  @exception std::out_of_range If idx is greater than the number of entry.
+       */
+      Entry getEntryByClusterOrder(entry_index_type idx) const;
+
+      /** Get the main entry of the archive.
+       *
+       *  @return The Main entry.
+       *  @exception EntryNotFound If no main entry has been specified in the archive.
+       */
+      Entry getMainEntry() const;
+
+      /** Check in an entry has path in the archive.
+       *
+       *  @param path The entry's path.
+       *  @return True if the path in the archive, false else.
+       */
+      bool hasEntryByPath(const std::string& path) const {
+        try{
+          getEntryByPath(path);
+          return true;
+        } catch(...) { return false; }
```

Comment:
> Avoid this kind of catch. It can be hiding some other problems.

I was a bit quick on this. We should use `impl->findx()` and check for the first entry if we have an exact match. No need to get a entry to catch the exception.

> In fact: Do we really need hasEntryBy*?

I've created them when porting the search implementation (see this discussion https://github.com/openzim/libzim/pull/378#discussion_r459984977).
But now the search directly use the fileImpl.
Having a `hasEntryBy` could be simpler that try/catch. But my first thought was to not implement them.